### PR TITLE
Set up Cobalt build acceleration with RBE

### PR DESCRIPTION
**Pull request by briantting**
b/384982606

### CODE REVIEW

File: DEPS [hunk 1]
```diff
@@ -261,7 +261,7 @@ vars = {
   # Fetch configuration files required for the 'use_remoteexec' gn arg
   'download_remoteexec_cfg': False,
   # RBE instance to use for running remote builds
-  'rbe_instance': Str('projects/rbe-chrome-untrusted/instances/default_instance'),
+  'rbe_instance': Str('projects/cobalt-actions-prod/instances/default_instance'),
```
Comment (andrewsavage1) [inline DEPS:264]: Can't this be specified in the .gclient file instead of us needing to change the DEPS file? Though I don't know if there's a good way to easily do this for all devs

File: buildtools/reclient_cfgs/rewrapper_linux_cc.cfg
Comment (andrewsavage1) [inline buildtools/reclient_cfgs/rewrapper_linux_cc.cfg:1]: We should probably put these files under //cobalt instead

File: build/toolchain/rbe.gni
Comment (andrewsavage1) [inline build/toolchain/rbe.gni:1]: Maybe there's a way we can set these variables for builds, but it would rely on gn.py—if someone runs `gn gen` locally without arguments do we care if they end up using the chromium reclient server vs ours?

File: DEPS (hunk 1 above)
Comment (briantting) [reply DEPS:264]: It should be able to be specified in `.gclient` much like how `download_remoteexec_cfg` is when we do `gclient config --name=src https://github.com/${{ github.repository }} --custom-var=download_remoteexec_cfg=True`, but I wasn't sure we wanted to ever "risk" using `rbe-chrome-untrusted`.

Also secondly, it's a dumb question but how do you pass "/" chars in key=value command line arg. `--custom-var=rbe_instance=projects/cobalt-actions-prod/instances/default_instance' throws errors. I've tried using quotes and escape chars

File: build/toolchain/rbe.gni
Comment (briantting) [reply build/toolchain/rbe.gni:1]: I rather not risk it.

1. they won't be able to because the default files will not exist since we aren't running CIPD to download them / we won't require people to have chromium CIPD access

2. if they did manage it would be slower since the Chromium's instances cache wont be updated with our compile tasks constantly like our instance's cache would be.

File: buildtools/reclient_cfgs/rewrapper_linux_cc.cfg
Comment (briantting) [reply buildtools/reclient_cfgs/rewrapper_linux_cc.cfg:1]: will do

File: DEPS (hunk 1 above)
Comment (andrewsavage1) [reply DEPS:264]: Seems like a quirk of gclient, not bash, but both of these worked for me:
```
$ gclient config --name=src https://github.com/youtube/cobalt.git --custom-var='download_remoteexec_cfg="foo/bar"'
$ gclient config --name=src https://github.com/youtube/cobalt.git --custom-var=download_remoteexec_cfg=\"foo/bar\"
```

File: build/toolchain/rbe.gni
Comment (andrewsavage1) [reply build/toolchain/rbe.gni:1]: Well people will have to use gn.py to build Cobalt (for now), and local devs are almost never building chromium, so it would only be relevant when people both (1) use gn gen with default args and then (2) change the gn gen args to set useremote_exec=true

Comment (briantting) [reply build/toolchain/rbe.gni:1]: I'm not sure I follow here. Why would we need the rbe cfg files to be variables? We will only have 1 config needed afaik.

Chrome has a rbe cfg file per rbe instance and manages it through CIPD downloads.

Comment (andrewsavage1) [reply build/toolchain/rbe.gni:1]: Just to avoid changing upstream code and changing the default way things build—it's a nice-to-have, not a must

File: DEPS (hunk 1 above)
Comment (briantting) [reply DEPS:264]: done!

File: build/toolchain/rbe.gni
Comment (briantting) [reply build/toolchain/rbe.gni:1]: Oh, I see what you mean. If someone tries to run without our specific

`gclient config --name=src git@github.com:youtube/cobalt.git --custom-var=download_remoteexec_cfg=True --custom-var='rbe_instance="projects/cobalt-actions-prod/instances/default_instance"'`

then they won't be able to use Chromium's rbe instance since this is a destructive change.

Comment (briantting) [reply build/toolchain/rbe.gni:1]: I think that's fine. No one should be using / be allowed to use Chromium's RBE instance anyways and even if we somehow preserved the cfg file paths here in `rbe.gni` with some variable it still wouldn't work due to the changes made in `fetch_reclient_cfgs.py` where we basically short circuit any CIPD download logic. No one should have CIPD download permissions either nor should they have Chromium's private docker image download permissions (private docker images are specified in the downloaded cfg files).

In short, we do not support the use of Chromium's RBE instance nor their cfg files nor their CIPD directory nor their private docker images. It's a cascade of permission requirements and no one should need to use their RBE instance over our RBE instance.

If by chance someone forgets our `gclient config` command's custom vars that's not a problem either since no build acceleration will be used at all. It will not fallback to Chromium's and cause a problem.

File: build/toolchain/rbe.gni [hunk 2]
```diff
@@ -18,17 +18,15 @@ declare_args() {
...
 declare_args() {
   # Set to the path of the RBE reclient configuration files.
   # Configuration file selection based on operating system.
   if (is_linux || is_android || is_chromeos || is_fuchsia) {
-    rbe_py_cfg_file = "${rbe_cfg_dir}/python/rewrapper_linux.cfg"
-    rbe_cc_cfg_file =
-        "${rbe_cfg_dir}/chromium-browser-clang/rewrapper_linux.cfg"
-    rbe_link_cfg_file =
-        "${rbe_cfg_dir}/chromium-browser-clang/rewrapper_linux_link.cfg"
+    rbe_py_cfg_file = "${cobalt_rbe_cfg_dir}/rewrapper_linux_py.cfg"
```
Comment (kaidokert) [inline build/toolchain/rbe.gni:28]: I'd recommend leaving this part intact, actually.

Create a `cobalt/infra/reclient_cfgs/` and move your config files under identical subdirectories ( python / chromium-browser-clang ) there. Then you don't need this modification here at all.

File: build/toolchain/rbe.gni [hunk 3]
```diff
@@ -18,17 +18,15 @@ declare_args() {
 
   # The directory where the re-client configuration files are.
   rbe_cfg_dir = rebase_path("//buildtools/reclient_cfgs", root_build_dir)
+  cobalt_rbe_cfg_dir = rebase_path("//cobalt/reclient_cfgs", root_build_dir)
```
Comment (kaidokert) [inline build/toolchain/rbe.gni:21]: I think you should be able to pass in `rbe_cfg_dir` as a variable to GN invocation / args.gn actually, so we dont need another variable here.

File: buildtools/reclient_cfgs/fetch_reclient_cfgs.py [hunk 4]
```diff
@@ -132,8 +135,9 @@ def main():
             logging.error(
               '--rbe_instance is required if --reproxy_cfg_template is set')
             return 1
-        if not GenerateReproxyCfg(args.reproxy_cfg_template, args.rbe_instance):
-           return 1
+        if GenerateReproxyCfg(args.reproxy_cfg_template, args.rbe_instance):
+           return 0
```
Comment (kaidokert) [inline buildtools/reclient_cfgs/fetch_reclient_cfgs.py:139]: Are the `return 0` and the extra `return 1` below correct ? Was this bug in upstream code ?

File: buildtools/reclient_cfgs/fetch_reclient_cfgs.py [hunk 5]
```diff
@@ -48,14 +48,17 @@ class CipdError(Exception):
 def CipdEnsure(pkg_name, ref, directory, quiet):
     logging.info('ensure %s %s in %s' % (pkg_name, ref, directory))
     log_level = 'warning' if quiet else 'debug'
+    auth = []
```
Comment (kaidokert) [inline buildtools/reclient_cfgs/fetch_reclient_cfgs.py:51]: It's not immediately clear and obvious that this modification in upstream code is Cobalt specific, so i would leave at least a comment saying this. It makes next rebase/update slightly easier.

**Review by kaidokert — COMMENTED** [review]
Took a pass through - please take a look.

Other than the comments, it looks okay

File: .github/actions/build/action.yaml
Comment (dahlstrom-g) [inline .github/actions/build/action.yaml:1]: It would be better to do unrelated modifications
in a separate commit with a title describing them.
It's a bit worse yet with related and unrelated
diff lines interspersed as they are here.

File: cobalt/build/gn.py [hunk 6]
```diff
@@ -85,7 +85,7 @@ def write_build_args(build_args_path, original_lines, dict_settings,
         f'The following args cannot be set in configs: {controlled_args}')
   gen_comment = '# Set by gn.py'
   with open(build_args_path, 'w', encoding='utf-8') as f:
-    if os.getenv('IS_CI', default='0') == '1':
+    if os.getenv('USE_CC_WRAPPER', default='0') == '1':
```
Comment (dahlstrom-g) [inline cobalt/build/gn.py:88]: Isn't `USE_CC_WRAPPER` itself unused?
Can't we just delete lines 88–90?

**Review by dahlstrom-g — CHANGES_REQUESTED** [review]

File: cobalt/build/gn.py (hunk 6 above)
Comment (briantting) [reply cobalt/build/gn.py:88]: We can, but keeping these lines here allows someone to use sccache as a fallback just by setting the env var.

File: build/toolchain/rbe.gni (hunk 2 above)
Comment (briantting) [reply build/toolchain/rbe.gni:28]: done.

File: build/toolchain/rbe.gni (hunk 3 above)
Comment (briantting) [reply build/toolchain/rbe.gni:21]: done.

File: buildtools/reclient_cfgs/fetch_reclient_cfgs.py (hunk 5 above)
Comment (briantting) [reply buildtools/reclient_cfgs/fetch_reclient_cfgs.py:51]: I can actually remove these changes, it was needed to enable auth for CIPD but now we no longer use CIPD.

File: buildtools/reclient_cfgs/fetch_reclient_cfgs.py (hunk 4 above)
Comment (briantting) [reply buildtools/reclient_cfgs/fetch_reclient_cfgs.py]: Added COBALT comment block. Basically we don't want any additional code here to execute as we no longer use CIPD for rewrapper configs and do not plan on supporting CIPD / deal with auth.

File: .github/actions/build/action.yaml
Comment (briantting) [reply .github/actions/build/action.yaml:1]: done.

File: .github/actions/build/action.yaml [hunk 7]
```diff
@@ -10,30 +10,30 @@ runs:
       - name: Get depot tools
         run: git clone https://chromium.googlesource.com/chromium/tools/depot_tools.git
         shell: bash
-      - name: Add directory to PATH
+      - name: Add depot tools to PATH
```
Comment (dahlstrom-g) [inline .github/actions/build/action.yaml:13]: still unrelated

File: .github/actions/build/action.yaml [hunk 8]
```diff
@@ -10,30 +10,30 @@ runs:
...
+        run: gclient config --name=src https://github.com/${{ github.repository }} --custom-var=download_remoteexec_cfg=True --custom-var='rbe_instance="projects/cobalt-actions-prod/instances/default_instance"'
         shell: bash
       - name: Set target OS for Android
         if: ${{ contains(matrix.platform, 'android') }}
         run: |
           echo "target_os=['android']" >> .gclient
           gclient validate
         shell: bash
-      - name: Gclient sync
+      - name: gclient sync
```
Comment (dahlstrom-g) [inline .github/actions/build/action.yaml:25]: still unrelated

File: .github/actions/build/action.yaml [hunk 9]
```diff
@@ -10,30 +10,30 @@ runs:
...
         run: gclient sync -v --shallow --no-history -r ${{ github.sha }}
         shell: bash
-      - name: sccache check
-        run: sccache -s
+      - name: Disable Chromium client side build telemetry
+        run: build_telemetry opt-out
         shell: bash
-      - name: gn gen
+      - name: GN gen
```
Comment (dahlstrom-g) [inline .github/actions/build/action.yaml:31]: still unrelated

File: .github/actions/build/action.yaml [hunk 10]
```diff
@@ -10,30 +10,30 @@ runs:
...
         run: |
           cd src
           cobalt/build/gn.py -p ${{ matrix.platform }} -C ${{ matrix.config }}
         shell: bash
-      - name: Dump GN args
+      - name: List GN args
```
Comment (dahlstrom-g) [inline .github/actions/build/action.yaml:36]: still unrelated

File: .github/actions/build/action.yaml
Comment (dahlstrom-g) [reply .github/actions/build/action.yaml:1]: not done?

File: cobalt/build/gn.py (hunk 6 above)
Comment (dahlstrom-g) [reply cobalt/build/gn.py:88]: I suppose you mean locally, but users don't
have that option before this PR unless perhaps
they set `IS_CI`, which seems unlikely.

Is this easter egg worth the extra complexity?
Will we even know when it stops working?
I don't think so; I say delete it now. ✊

YAGNI and all that.

**Review by dahlstrom-g — COMMENTED** [review]
I wrote the comment below in my previous reply
but GitHub silently lost it, so I had to write it again.
This has happened to me repeatedly. 😠
GitHub PRs feel inadequate as a review tool.

File: .github/actions/build/action.yaml (hunk 7 above)
Comment (briantting) [reply .github/actions/build/action.yaml:13]: It's separated into distinct commits though or did you mean it should be in its own PR?

File: cobalt/build/gn.py (hunk 6 above)
Comment (briantting) [reply cobalt/build/gn.py:88]: done.

File: .github/actions/build/action.yaml
Comment (briantting) [reply .github/actions/build/action.yaml:1]: ack!

**Review by kaidokert — APPROVED** [review]
My comments are all in, LGTM :+1:

**Review by dahlstrom-g — DISMISSED** [review]
LGTM (especially if not squashed)
… are the commits in the right order? 🤷

File: .github/actions/build/action.yaml [hunk 11]
```diff
@@ -10,30 +10,30 @@ runs:
       - name: Get depot tools
         run: git clone https://chromium.googlesource.com/chromium/tools/depot_tools.git
         shell: bash
-      - name: Add directory to PATH
+      - name: Add depot tools to PATH
         run: echo "$GITHUB_WORKSPACE/depot_tools" >> $GITHUB_PATH
         shell: bash
       - name: Generate gclient file
-        run: gclient config --name=src https://github.com/${{ github.repository }}
+        run: gclient config --name=src https://github.com/${{ github.repository }} --custom-var=download_remoteexec_cfg=True --custom-var='rbe_instance="projects/cobalt-actions-prod/instances/default_instance"'
```
Comment (dahlstrom-g) [inline .github/actions/build/action.yaml:17]: nit: Options are prescribed _before_ the URL:
`Usage: gclient.py config [options] [url]`

**Review by dahlstrom-g — COMMENTED** [review]
As it stands I believe this PR will break Kokoro builds.

nits:
The word “Chrobalt” should probably be
dropped from the PR and commit titles.
The PR and commit titles shouldn't have periods.
https://cbea.ms/git-commit/#end
Each commit title should be imperative.
https://cbea.ms/git-commit/#imperative
